### PR TITLE
concurrent_unordered_map updates

### PR DESCRIPTION
**Pull request by jrhemstad**
New functionality:
- [x] Non-aggregating `insert` function
- [x] Specialize `insert` for when the key/value update can be done with a single CAS
- [x] Tests for new insert

Maintenance
- [x] Default to `legacy_allocator`
- [x] Moved `unused_key` to member variable instead of template arg
- [x] Renamed `m_hashtbl_size` and `m_hashtbl_capacity` to just `capacity` as they were being used interchangeably and referring to "size" was misleading 
- [x] Simplified `find` implementation

### CODE REVIEW

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 1]
```diff
@@ -17,37 +17,35 @@
 #ifndef CONCURRENT_UNORDERED_MAP_CUH
 #define CONCURRENT_UNORDERED_MAP_CUH
 
-#include <iterator>
-#include <type_traits>
-#include <cassert>
-#include <iostream>
-#include <thrust/pair.h>
 
 #include "cudf.h"
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Use `#include <>` except for files in same directory.

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 2]
```diff
@@ -17,37 +17,34 @@
...
-#include <iterator>
-#include <type_traits>
-#include <cassert>
-#include <iostream>
-#include <thrust/pair.h>
-
-#include "cudf.h"
-#include "groupby/aggregation_operations.hpp"
-
+#include <cudf.h>
```
Comment (nsakharnykh) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Do all the cuDF includes here necessary? I'd like to see if we can avoid dependencies on cuDF as much as possible, since this class may be used outside cuDF for other applications.

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 3]
```diff
@@ -87,50 +99,39 @@ public:
...
-    static constexpr __host__ __device__ key_type get_unused_key()
-    {
-        return unused_key;
+    __device__ const_iterator end() const {
+      return const_iterator(m_hashtbl_values, m_hashtbl_values + capacity,
+                            m_hashtbl_values + capacity);
     }
+    __device__ value_type* data() const { return m_hashtbl_values; }
+
+    __host__ __device__ key_type get_unused_key() const { return m_unused_key; }
```
Comment (nsakharnykh) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Why did you change this from `constexpr` to `const`?

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 4]
```diff
@@ -57,28 +54,43 @@ public:
...
+   __device__
+   pair_packer(value_type _pair) : pair{_pair} {}
+
+   __device__
+   pair_packer(uint64_t _packed) : packed{_packed} {}
+ };
+
 public:
 
     explicit concurrent_unordered_map(size_type n,
```
Comment (nsakharnykh) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Should add documentation for the constructor

File: cpp/src/hash/concurrent_unordered_map.cuh (hunk 3 above)
Comment (jrhemstad) [reply cpp/src/hash/concurrent_unordered_map.cuh]: Because `unused_key` is no longer specified via template parameter, but rather via a parameter of the classes constructor.

File: cpp/src/hash/concurrent_unordered_map.cuh (hunk 2 above)
Comment (jrhemstad) [reply cpp/src/hash/concurrent_unordered_map.cuh]: You're right, these were unnecessary.

File: CHANGELOG.md [hunk 5]
```diff
@@ -10,12 +10,14 @@
 
 ## Improvements
 
+- PR #1703 Added non-aggregating `insert` to `concurrent_unordered_map` with specializations to store pairs with a single atomicCAS when possible.
```
Comment (harrism) [inline CHANGELOG.md]:
```suggestion
- PR #1703 C++: Added non-aggregating `insert` to `concurrent_unordered_map` with specializations to store pairs with a single atomicCAS when possible.
```

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 6]
```diff
@@ -282,137 +336,183 @@ public:
...
+     * hash bucket
+     *---------------------------------------------------------------------------**/
+    enum class insert_result {
+      CONTINUE,  ///< Insert did not succeed, continue trying to insert (collision)
+      SUCCESS,   ///< New pair inserted successfully
+      DUPLICATE  ///< Insert did not succeed, key is already present
+    };
+
+    /**---------------------------------------------------------------------------*
+     * @brief Specialization for value types that can be packed.
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]: I think there's an @overload oxygen tag.

**Comment by jrhemstad** [comment]
@nsakharnykh can you re-review so that this can be merged?

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 7]
```diff
@@ -57,84 +104,98 @@ public:
...
+       CUDA_RT_CALL(cudaGetDevice(&dev_id));
+       CUDA_RT_CALL(cudaMemPrefetchAsync(
+           m_hashtbl_values, m_capacity * sizeof(value_type), dev_id, 0));
+     }
+   }
+
+   init_hashtbl<<<((m_capacity - 1) / block_size) + 1, block_size>>>(
+       m_hashtbl_values, m_capacity, m_unused_key, m_unused_element);
+   CUDA_RT_CALL(cudaGetLastError());
+   CUDA_RT_CALL(cudaStreamSynchronize(0));
```
Comment (nsakharnykh) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Do we want to remove this and add an optional stream parameter to the constructor? Given that we're moving to async model for other operators and have a chance to fix it here for the concurrent hash map.

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 8]
```diff
@@ -282,137 +342,184 @@ public:
...
+     * Returns an iterator, boolean pair.
+     *
+     * If the new key already present in the map, the iterator points to
+     * the location of the existing key and the boolean is `false` indicating
+     * that the insert did not succeed.
+     *
+     * If the new key was not present, the iterator points to the location where
+     * the insert occured and the boolean is `true` indicating that the insert
+     * succeeded.
+     *
```
Comment (nsakharnykh) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Should add the same @note that `insert` is not threadsafe with `find`

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 9]
```diff
@@ -282,137 +342,184 @@ public:
...
+        if (m_equal(m_unused_key, existing_key)) {
+          return this->end();
+        }
+        index = (index + 1) % m_capacity;
+        current_bucket = &m_hashtbl_values[index];
+      }
     }
-    
+
     gdf_error assign_async( const concurrent_unordered_map& other, cudaStream_t stream = 0 )
```
Comment (nsakharnykh) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Brief description?

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 10]
```diff
@@ -282,137 +342,184 @@ public:
...
             
-            m_hashtbl_values = m_allocator.allocate( m_hashtbl_capacity );
+            m_hashtbl_values = m_allocator.allocate( m_capacity );
         }
-        CUDA_TRY( cudaMemcpyAsync( m_hashtbl_values, other.m_hashtbl_values, m_hashtbl_size*sizeof(value_type), cudaMemcpyDefault, stream ) );
+        CUDA_TRY( cudaMemcpyAsync( m_hashtbl_values, other.m_hashtbl_values, m_capacity*sizeof(value_type), cudaMemcpyDefault, stream ) );
         return GDF_SUCCESS;
     }
     
     void clear_async( cudaStream_t stream = 0 )
```
Comment (nsakharnykh) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Again, brief description?

File: cpp/src/hash/concurrent_unordered_map.cuh (hunk 7 above)
Comment (nsakharnykh) [reply cpp/src/hash/concurrent_unordered_map.cuh]: I saw the TODO, we can do it in a separate PR
